### PR TITLE
Make StringLiterals and HeritageClauses into TypeNodes. 

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Make StringLiterals and HeritageClauses into TypeNodes. It makes the checker much simpler and nicer.

### CODE REVIEW

**Comment by jbondc** [comment]
:smile:

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2289,18 +2289,18 @@ module ts {
             return links.type;
         }
 
-        function getSetAccessorTypeAnnotationNode(accessor: AccessorDeclaration): TypeNode | LiteralExpression {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: We allow string literals in a set accessor?

File: src/compiler/types.ts [hunk 2]
```diff
@@ -689,7 +689,7 @@ module ts {
         hasExtendedUnicodeEscape?: boolean;
     }
 
-    export interface StringLiteralExpression extends LiteralExpression {
+    export interface StringLiteralExpression extends LiteralExpression, TypeNode {
```
Comment (JsonFreeman) [inline src/compiler/types.ts]: Just a naming issue, but I feel like this should be StringLiteral more than StringLiteralExpression. If it's a type node, it's not an expression

**Comment by JsonFreeman** [comment]
:+1:

File: src/compiler/checker.ts (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: I'm not certain.  It's allowed by the grammar.  But i'm not sure it has real meaning.  Regardless, it doesn't need ot be explicitly stated like so.

File: src/compiler/types.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: Good point!

**Pull request merged**